### PR TITLE
Fixing performance issues and out-of-order packets

### PR DESCRIPTION
**Pull request by urbanishimwe**
### Reducing CPU context switching and number of goroutines. 
Packet capture and packet processing now use only two goroutines which helps to minimize CPU context switches. Spawning too many goroutines is harmful here. 

### Optimized packet capture - allocated memory only when required, and only for data which is used
Using ZeroCopy methods from libpcap library to avoid unnecessary allocations. Now memory gets allocated ONLY for the valid packets, and only for the packets which have the data. E.g. no SYN/FIN packets are used now. Additionally we now use `sync.Pool` for re-using packet objects, which helps to re-use already allocated memory. 

### Simplification and optimization of request/response detection
There is no SYN/FIN packets anymore etc. Now only packet payload is used to detect start and end of the packet. More over payload detection now does not require generating a total “message” buffer, and works with individual packet payloads. 

Message payloads now concatenated from packets only in the end when message is dispatched. Also, before checking if message is complete, added additional check if all received packets in the valid order, e.g. if their SEQ is valid, and no packets are missing. 

Reworked chunked encoding validation, and now it does not need expensive operation of re-calculating all the chunks. Now it “trust” that client gives valid chunk body, check if packets are in the right order (e.g. SEQ match), and checks if message ends with the right suffix. All is done with 0 allocations. 

Parsing all Headers using `proto.GetHeaders` was proved to be very slow. Now we only parse the headers we need(and do it only once).

Packets gets matched together using ACK, which on high RPS removed chances of duplicating IDs. Additionally, even if packets are received out of order, now it will properly sort them, before dispatching the message.

### Changes in ID generation algorithm
Message ID generation and relations between request and response IDs is fully rewritten. Responses now do not have to lookup for request data in order to get the same ID. ID no rely on the fact that SEQ of the first packet of the response should be the same as ACK of the request. If previously Message ID contained random values, like current timestamp, now it has a consistent algorithm which is based on TCP stream id (SrcPort + DstPort + SrcIP/DstIP) and current ACK/SEQ number (to distinguish multiple messages within the same stream).

### BPF filter optimizations
When tracking response it now uses a more accurate BPF rule to filter only needed traffic. 

### Misc
The packet code is now fully moved to tcp/Packet, so packet processing done only once in one place.

TCP output now has a 5 second timeout, and has a proper Close method.

Fully switching to go modules and removing vendoring.

### CODE REVIEW

**Comment by buger** [comment]
Since there are a lot of mutex usage now, maybe think about using https://github.com/cornelk/hashmap instead.

**Comment by buger** [comment]
The main thing to ensure here is that packet "capture" happens in one thread, and packet "processing" in another. 
And actually if we ensure that all packet processing is inside "single" goroutine, then we do not need to put mutexes around maps. And "slower" performance processing goroutine will not affect that much "capture" goroutine.

**Review by urbanishimwe — COMMENTED** [review]
this naming makes more sense, thanks!

File: proto/proto.go [hunk 1]
```diff
@@ -432,87 +432,87 @@ func CheckChunked(buf []byte) (chunkEnd int, full bool) {
 	return
 }
 
-// Message is an interface used to provide feedback or store dummy data for future use
-type Message interface {
+// Message is an interface used to provide protocol state or store dummy data for future use
+type ProtocolStateSetter interface {
```
Comment (urbanishimwe) [inline proto/proto.go:436]: Since this interface has both setter(`SetProtocolState`) and getter(`ProtocolState`), it may be called `ProtocolState`

**Pull request merged**